clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/ActiveActiveStoreIngestionTask.java

```diff
@@ -491,30 +494,65 @@ protected void processMessageAndMaybeProduceToKafka(
       // call in this context much less obtrusive, however, it implies that all views can only work for AA stores
       int valueSchemaId =
           rmdWithValueSchemaID != null ? rmdWithValueSchemaID.getValueSchemaId() : incomingValueSchemaId;
-      this.viewWriters.forEach(
-          (k, v) -> v.processRecord(
-              mergeConflictResult.getNewValue(),
-              oldValueProvider.get(),
-              keyBytes,
-              versionNumber,
-              incomingValueSchemaId,
-              valueSchemaId,
-              mergeConflictResult.getRmdRecord()));
 
-      // This function may modify the original record in KME and it is unsafe to use the payload from KME directly after
-      // this call.
-      producePutOrDeleteToKafka(
-          mergeConflictResult,
-          partitionConsumptionState,
-          keyBytes,
-          consumerRecord,
-          subPartition,
-          kafkaUrl,
-          kafkaClusterId,
-          beforeProcessingRecordTimestampNs,
-          valueManifestContainer.getManifest(),
-          rmdWithValueSchemaID == null ? null : rmdWithValueSchemaID.getRmdManifest());
+      // Get Futures and
+      // Pass callback with on completion which decrements, once all decrement call producePutOrDeleteToKafka
+      if (this.viewWriters.size() > 0) {
+        Long preprocessingTime = System.currentTimeMillis();
+
+        List<Future> viewWriteCallbacks = new ArrayList<>();
+        for (VeniceViewWriter writer: viewWriters.values()) {
+          viewWriteCallbacks.add(
+              writer.processRecord(
+                  mergeConflictResult.getNewValue(),
+                  oldValueProvider.get(),
+                  keyBytes,
+                  versionNumber,
+                  incomingValueSchemaId,
+                  valueSchemaId,
+                  mergeConflictResult.getRmdRecord()));
+        }
+        viewWriteCallbacks.add(partitionConsumptionState.getLastVTProduceCallFuture());
+        CompletableFuture.allOf(viewWriteCallbacks.toArray(new CompletableFuture[0]))
+            .whenCompleteAsync((value, exception) -> {
+              hostLevelIngestionStats.recordViewProducerLatency(LatencyUtils.getLatencyInMS(preprocessingTime));
+              if (exception == null) {
+                partitionConsumptionState.setLastVTProduceCallFuture(CompletableFuture.runAsync(() -> {
```

Comment:
Hmm... IDK if the design we discussed works if this CF is declared inside the `whenCompleteAsync` 🤔 ...

There could be many writes in a row that use the same "previous VT write future" as we're waiting for this `whenCompleteAsync` closure to kick in, no?

Reply:
I think this works, but I think theres one last bit of tedium here.  If we set the future stored on the PCS as one which has completed exceptionally, how do we clear it?  If we take the last one and it had an exception and then set the stored future as a 'allOf' then I think it'll be exceptional every time?  I think we'll hit a doom loop.  We need a mechanism to clear it I think....

Reply:
Maybe at the top check if it already completed exceptionally and bubble up?

Reply:
Hmm... yeah, that's a doom loop if I've ever seen one. If we fail to write to VT, then I guess it's not right to move on... we should either stall and retry, or fail outright. And either way, we should record that we failed via a metric... today we have error replica metrics, but after this swallowing  future refactor, I guess that metric won't have a chance of firing... We should find a way to bubble it up, somehow, but I'm not sure what is a good code structure to get there.

Reply:
`          } else {
            VeniceException veniceException = new VeniceException(exception);
            this.setIngestionException(partitionConsumptionState.getPartition(), veniceException);
            currentVersionTopicWrite.completeExceptionally(veniceException);
          }`

Reply:
Sure, but if we call `currentVersionTopicWrite.completeExceptionally` then we're in a doom loop, right?

Reply:
We will be in a doom loop for a while, but my understanding of the ingestionException is that if theres an exception in there the task will fail eventually...  I think.

Reply:
SG.